File: binding/src/libqscrollarea.cpp
#include <QAbstractScrollArea>
#include <QEvent>
#include <QObject>
#include <QResizeEvent>
#include <QScrollArea>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWidget>
#include <qscrollarea.h>
#include "libqscrollarea.h"
#include "libqscrollarea.hxx"

QScrollArea* QScrollArea_new(QWidget* parent) {
	 return new VirtualQScrollArea(parent);
}

QScrollArea* QScrollArea_new2() {
	 return new VirtualQScrollArea();
}

libqt_string QScrollArea_Tr(const char* s) {
	QString _ret = QScrollArea::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QWidget* QScrollArea_Widget(const QScrollArea* self) {
	return self->widget();
}

void QScrollArea_SetWidget(QScrollArea* self, QWidget* widget) {
	self->setWidget(widget);
}

QWidget* QScrollArea_TakeWidget(QScrollArea* self) {
	return self->takeWidget();
}

bool QScrollArea_WidgetResizable(const QScrollArea* self) {
	return self->widgetResizable();
}

void QScrollArea_SetWidgetResizable(QScrollArea* self, bool resizable) {
	self->setWidgetResizable(resizable);
}

QSize* QScrollArea_SizeHint(const QScrollArea* self) {
	return new QSize(self->sizeHint());
}

bool QScrollArea_FocusNextPrevChild(QScrollArea* self, bool next) {
	return self->focusNextPrevChild(next);
}

int QScrollArea_Alignment(const QScrollArea* self) {
	return self->alignment();
}

void QScrollArea_SetAlignment(QScrollArea* self, int alignment) {
	self->setAlignment(static_cast<QFlags<Qt::AlignmentFlag>>(alignment));
}

void QScrollArea_EnsureVisible(QScrollArea* self, int x, int y) {
	self->ensureVisible(x, y);
}

void QScrollArea_EnsureWidgetVisible(QScrollArea* self, QWidget* childWidget) {
	self->ensureWidgetVisible(childWidget);
}

libqt_string QScrollArea_Tr2(const char* s, const char* c) {
	QString _ret = QScrollArea::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QScrollArea_Tr3(const char* s, const char* c, int n) {
	QString _ret = QScrollArea::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QScrollArea_EnsureVisible3(QScrollArea* self, int x, int y, int xmargin) {
	self->ensureVisible(x, y, xmargin);
}

void QScrollArea_EnsureVisible4(QScrollArea* self, int x, int y, int xmargin, int ymargin) {
	self->ensureVisible(x, y, xmargin, ymargin);
}

void QScrollArea_EnsureWidgetVisible2(QScrollArea* self, QWidget* childWidget, int xmargin) {
	self->ensureWidgetVisible(childWidget, xmargin);
}

void QScrollArea_EnsureWidgetVisible3(QScrollArea* self, QWidget* childWidget, int xmargin, int ymargin) {
	self->ensureWidgetVisible(childWidget, xmargin, ymargin);
}

// Base class handler implementation
QSize* QScrollArea_QBaseSizeHint(const QScrollArea* self) {
	auto* vqscrollarea = dynamic_cast<const VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_SizeHint_IsBase(true);
	return new QSize(vqscrollarea->sizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnSizeHint(const QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<const VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_SizeHint_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_SizeHint_Callback>(slot));
}
}

// Base class handler implementation
bool QScrollArea_QBaseFocusNextPrevChild(QScrollArea* self, bool next) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_FocusNextPrevChild_IsBase(true);
	return vqscrollarea->focusNextPrevChild(next);
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnFocusNextPrevChild(QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_FocusNextPrevChild_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_FocusNextPrevChild_Callback>(slot));
}
}

// Derived class handler implementation
bool QScrollArea_Event(QScrollArea* self, QEvent* param1) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
	return vqscrollarea->event(param1);
	} else {
	return self->QScrollArea::event(param1);
}
}

// Base class handler implementation
bool QScrollArea_QBaseEvent(QScrollArea* self, QEvent* param1) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_Event_IsBase(true);
	return vqscrollarea->event(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnEvent(QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_Event_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_Event_Callback>(slot));
}
}

// Derived class handler implementation
bool QScrollArea_EventFilter(QScrollArea* self, QObject* param1, QEvent* param2) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
	return vqscrollarea->eventFilter(param1, param2);
	} else {
	return self->QScrollArea::eventFilter(param1, param2);
}
}

// Base class handler implementation
bool QScrollArea_QBaseEventFilter(QScrollArea* self, QObject* param1, QEvent* param2) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_EventFilter_IsBase(true);
	return vqscrollarea->eventFilter(param1, param2);
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnEventFilter(QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_EventFilter_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_EventFilter_Callback>(slot));
}
}

// Derived class handler implementation
void QScrollArea_ResizeEvent(QScrollArea* self, QResizeEvent* param1) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
	vqscrollarea->resizeEvent(param1);
	} else {
	self->QScrollArea::resizeEvent(param1);
}
}

// Base class handler implementation
void QScrollArea_QBaseResizeEvent(QScrollArea* self, QResizeEvent* param1) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_ResizeEvent_IsBase(true);
	vqscrollarea->resizeEvent(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnResizeEvent(QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_ResizeEvent_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_ResizeEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QScrollArea_ScrollContentsBy(QScrollArea* self, int dx, int dy) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
	vqscrollarea->scrollContentsBy(dx, dy);
	} else {
	self->QScrollArea::scrollContentsBy(dx, dy);
}
}

// Base class handler implementation
void QScrollArea_QBaseScrollContentsBy(QScrollArea* self, int dx, int dy) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_ScrollContentsBy_IsBase(true);
	vqscrollarea->scrollContentsBy(dx, dy);
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnScrollContentsBy(QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_ScrollContentsBy_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_ScrollContentsBy_Callback>(slot));
}
}

// Derived class handler implementation
QSize* QScrollArea_ViewportSizeHint(const QScrollArea* self) {
	auto* vqscrollarea = dynamic_cast<const VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
	return new QSize(vqscrollarea->viewportSizeHint());
	} else {
	return new QSize(self->QScrollArea::viewportSizeHint());
}
}

// Base class handler implementation
QSize* QScrollArea_QBaseViewportSizeHint(const QScrollArea* self) {
	auto* vqscrollarea = dynamic_cast<const VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_ViewportSizeHint_IsBase(true);
	return new QSize(vqscrollarea->viewportSizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QScrollArea_OnViewportSizeHint(const QScrollArea* self, intptr_t slot) {
	auto* vqscrollarea = dynamic_cast<const VirtualQScrollArea*>(self);
	if (vqscrollarea && vqscrollarea->isVirtualQScrollArea) {
vqscrollarea->setQScrollArea_ViewportSizeHint_Callback(reinterpret_cast<VirtualQScrollArea::QScrollArea_ViewportSizeHint_Callback>(slot));
}
}

void QScrollArea_Delete(QScrollArea* self) {
    delete self;
}

File: binding/include/libqscrollarea.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qscrollarea.html
pub const qscrollarea = struct {

    /// New constructs a new QScrollArea object.
    pub fn New(parent: ?*anyopaque) QtC.QScrollArea {
        return qtc.QScrollArea_new(@ptrCast(parent));
    }


    /// New2 constructs a new QScrollArea object.
    pub fn New2() QtC.QScrollArea {
        return qtc.QScrollArea_new2();
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScrollArea_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscrollarea.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Widget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QScrollArea_Widget(@ptrCast(self));
    }

    pub fn SetWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QScrollArea_SetWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn TakeWidget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QScrollArea_TakeWidget(@ptrCast(self));
    }

    pub fn WidgetResizable(self: ?*anyopaque, ) bool {
        return qtc.QScrollArea_WidgetResizable(@ptrCast(self));
    }

    pub fn SetWidgetResizable(self: ?*anyopaque, resizable: bool) void {
        qtc.QScrollArea_SetWidgetResizable(@ptrCast(self), resizable);
    }

    pub fn SizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QScrollArea_SizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QScrollArea_OnSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QScrollArea_QBaseSizeHint(@ptrCast(self));
    }

    pub fn FocusNextPrevChild(self: ?*anyopaque, next: bool) bool {
        return qtc.QScrollArea_FocusNextPrevChild(@ptrCast(self), next);
    }

    /// Allows for overriding the related default method
    pub fn OnFocusNextPrevChild(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) bool) void {
        qtc.QScrollArea_OnFocusNextPrevChild(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFocusNextPrevChild(self: ?*anyopaque, next: bool) bool {
        return qtc.QScrollArea_QBaseFocusNextPrevChild(@ptrCast(self), next);
    }

    pub fn Alignment(self: ?*anyopaque, ) i32 {
        return qtc.QScrollArea_Alignment(@ptrCast(self));
    }

    pub fn SetAlignment(self: ?*anyopaque, alignment: i32) void {
        qtc.QScrollArea_SetAlignment(@ptrCast(self), @intCast(alignment));
    }

    pub fn EnsureVisible(self: ?*anyopaque, x: i32, y: i32) void {
        qtc.QScrollArea_EnsureVisible(@ptrCast(self), x, y);
    }

    pub fn EnsureWidgetVisible(self: ?*anyopaque, childWidget: ?*anyopaque) void {
        qtc.QScrollArea_EnsureWidgetVisible(@ptrCast(self), @ptrCast(childWidget));
    }

    pub fn Event(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QScrollArea_Event(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QScrollArea_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QScrollArea_QBaseEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn EventFilter(self: ?*anyopaque, param1: ?*anyopaque, param2: ?*anyopaque) bool {
        return qtc.QScrollArea_EventFilter(@ptrCast(self), @ptrCast(param1), @ptrCast(param2));
    }

    /// Allows for overriding the related default method
    pub fn OnEventFilter(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QScrollArea_OnEventFilter(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEventFilter(self: ?*anyopaque, param1: ?*anyopaque, param2: ?*anyopaque) bool {
        return qtc.QScrollArea_QBaseEventFilter(@ptrCast(self), @ptrCast(param1), @ptrCast(param2));
    }

    pub fn ResizeEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QScrollArea_ResizeEvent(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnResizeEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QScrollArea_OnResizeEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseResizeEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QScrollArea_QBaseResizeEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn ScrollContentsBy(self: ?*anyopaque, dx: i32, dy: i32) void {
        qtc.QScrollArea_ScrollContentsBy(@ptrCast(self), dx, dy);
    }

    /// Allows for overriding the related default method
    pub fn OnScrollContentsBy(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, i32) callconv(.c) void) void {
        qtc.QScrollArea_OnScrollContentsBy(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseScrollContentsBy(self: ?*anyopaque, dx: i32, dy: i32) void {
        qtc.QScrollArea_QBaseScrollContentsBy(@ptrCast(self), dx, dy);
    }

    pub fn ViewportSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QScrollArea_ViewportSizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnViewportSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QScrollArea_OnViewportSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseViewportSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QScrollArea_QBaseViewportSizeHint(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScrollArea_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscrollarea.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScrollArea_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscrollarea.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn EnsureVisible3(self: ?*anyopaque, x: i32, y: i32, xmargin: i32) void {
        qtc.QScrollArea_EnsureVisible3(@ptrCast(self), x, y, xmargin);
    }

    pub fn EnsureVisible4(self: ?*anyopaque, x: i32, y: i32, xmargin: i32, ymargin: i32) void {
        qtc.QScrollArea_EnsureVisible4(@ptrCast(self), x, y, xmargin, ymargin);
    }

    pub fn EnsureWidgetVisible2(self: ?*anyopaque, childWidget: ?*anyopaque, xmargin: i32) void {
        qtc.QScrollArea_EnsureWidgetVisible2(@ptrCast(self), @ptrCast(childWidget), xmargin);
    }

    pub fn EnsureWidgetVisible3(self: ?*anyopaque, childWidget: ?*anyopaque, xmargin: i32, ymargin: i32) void {
        qtc.QScrollArea_EnsureWidgetVisible3(@ptrCast(self), @ptrCast(childWidget), xmargin, ymargin);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QScrollArea_Delete(@ptrCast(self));
    }
};
